src/envoy/extensions/wasm/wasm.cc

```diff
@@ -118,22 +136,94 @@ WasmHandleExtensionCloneFactory IstioWasmExtension::wasmCloneFactory() {
   };
 }
 
+static std::string statsKey(const PluginSharedPtr& plugin) {
+  auto sep = std::string("\t");
+  return plugin->name_ + sep + plugin->runtime_;
+}
+
 void IstioWasmExtension::onEvent(WasmEvent event,
                                  const PluginSharedPtr& plugin) {
   EnvoyWasm::onEvent(event, plugin);
+  auto key = statsKey(plugin);
+  auto& stats = config_stats_.at(key);
+  switch (event) {
+    case EnvoyWasm::WasmEvent::Ok:
+    case EnvoyWasm::WasmEvent::RemoteLoadCacheHit:
+      break;
+    case EnvoyWasm::WasmEvent::RemoteLoadCacheNegativeHit:
+      stats->permanent_read_error_->inc();
+      break;
+    case EnvoyWasm::WasmEvent::RemoteLoadCacheMiss:
+      stats->eventually_consistent_read_->inc();
+      break;
+    case EnvoyWasm::WasmEvent::RemoteLoadCacheFetchSuccess:
+      break;
+    case EnvoyWasm::WasmEvent::RemoteLoadCacheFetchFailure:
+      stats->permanent_read_error_->inc();
+      break;
+    case EnvoyWasm::WasmEvent::UnableToCreateVM:
+    case EnvoyWasm::WasmEvent::UnableToCloneVM:
+    case EnvoyWasm::WasmEvent::MissingFunction:
+    case EnvoyWasm::WasmEvent::UnableToInitializeCode:
+      stats->invalid_module_->inc();
+      break;
+    case EnvoyWasm::WasmEvent::StartFailed:
+    case EnvoyWasm::WasmEvent::ConfigureFailed:
+      stats->invalid_configuration_->inc();
+      break;
+    case EnvoyWasm::WasmEvent::RuntimeError:
+      break;
+  }
 }
 
 void IstioWasmExtension::onRemoteCacheEntriesChanged(int entries) {
   EnvoyWasm::onRemoteCacheEntriesChanged(entries);
-  create_wasm_stats_->remote_load_cache_entries_.set(entries);
 }
 
+// NB: the "scope" here is tied to the lifetime of the filter chain in many
+// cases and may disappear. Code in envoy detects that and will call
+// resetStats().
 void IstioWasmExtension::createStats(const Stats::ScopeSharedPtr& scope,
                                      const PluginSharedPtr& plugin) {
   EnvoyWasm::createStats(scope, plugin);
+  auto istio_version = plugin->local_info_.node()
+                           .metadata()
+                           .fields()
+                           .at("ISTIO_VERSION")
```

Comment:
This seems to be causing segfault.

Reply:
yes, falling back to Envoy version when the ISTIO_VERSION is not set.